src/schema/integrations.ts

```diff
@@ -1,38 +1,191 @@
+export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers({
   Query: {
-    popularIntegrations: async (source, args, ctx): Promise<GQLIntegration[]> =>
-      ctx.getRepository(Integration).find({ order: { timestamp: 'DESC' } }),
+    slackChannels: async (
+      _,
+      args: { integrationId: string; limit: number; cursor: string },
+      ctx: AuthContext,
+    ): Promise<GQLSlackChannelConnection> => {
+      const slackIntegration = await getSlackIntegrationOrFail({
+        id: args.integrationId,
+        userId: ctx.userId,
+        con: ctx.con,
+      });
+
+      const client = new WebClient(
+        await getIntegrationToken({ integration: slackIntegration }),
+      );
+
+      const result = await client.conversations.list({
+        limit: getLimit({
+          limit: args.limit,
+          defaultLimit: 10,
+          max: 999,
+        }),
+        cursor: args.cursor,
+        exclude_archived: true,
+      });
+
+      if (!result.ok) {
+        const message = 'failed to fetch slack channels';
+
+        logger.error(
+          {
+            err: new Error(result.error),
+          },
+          message,
+        );
+
+        throw new Error(message);
+      }
+
+      return {
+        data: result.channels.map((channel) => {
+          return {
+            id: channel.id,
+            name: channel.name,
+          };
+        }),
+        cursor: result.response_metadata?.next_cursor,
+      };
+    },
+  },
+  Mutation: {
+    slackConnectSource: async (
+      _,
+      args: { integrationId: string; channelId: string; sourceId: string },
+      ctx: AuthContext,
+    ): Promise<GQLEmptyResponse> => {
+      const [slackIntegration, source] = await Promise.all([
+        getSlackIntegrationOrFail({
+          id: args.integrationId,
+          userId: ctx.userId,
+          con: ctx.con,
+        }),
+        ctx.con.getRepository(Source).findOneByOrFail({
+          id: args.sourceId,
+        }),
+      ]);
+
+      const client = new WebClient(
+        await getIntegrationToken({ integration: slackIntegration }),
+      );
+
+      const channelResult = await client.conversations.info({
+        channel: args.channelId,
+      });
+
+      if (!channelResult.ok && channelResult.channel.id !== args.channelId) {
+        throw new ValidationError('invalid channel');
+      }
+
+      await ctx.con.getRepository(UserSourceIntegrationSlack).upsert(
+        {
+          userIntegrationId: slackIntegration.id,
+          sourceId: args.sourceId,
+          // only one channel per source is allowed currently
+          channelIds: [args.channelId],
+        },
+        ['userIntegrationId', 'sourceId'],
+      );
+
+      await client.conversations.join({
+        channel: args.channelId,
+      });
+
+      await client.chat.postMessage({
+        channel: args.channelId,
+        text: `Connected source "${source.name}" to this channel 🙌`,
```

Comment:
Let's get the copy from product for this. Also I suggest adding an invitation link to the message

Reply:
I asked for copy but I'll merge this and update after since I'll do another PR for some stuff later.

Reply:
please remember to add also link in the next PR :)

Reply:
Updated the copy now. I'll see about the link in that next one yes when I sort out joining.